spdlog/src/formatter/json_formatter.rs

```diff
@@ -0,0 +1,258 @@
+/// | `payload`   | String       | The contents of the log.                                                                                                       |
+/// | `logger`    | String/Null  | The name of the logger. Null if the logger has no name.                                                                        |
+/// | `tid`       | Integer(u64) | The thread ID when the log was generated.                                                                                      |
+/// | `source`    | Object/Null  | The source location of the log. See [`SourceLocation`] for its schema. Null if crate feature `source-location` is not enabled. |
+/// 
+/// <div class="warning">
+/// 
+/// - If the type of a field is Null, the field will not be present or be `null`.
+/// 
+/// - The order of the fields is not guaranteed.
+/// 
+/// </div>
+/// 
+/// ---
+/// 
+/// ## Examples
+///
+///  - Default:
+/// 
+///    ```json
+///    {"level":"Info","timestamp":1722817424798,"payload":"hello, world!","tid":3472525}
+///    {"level":"Error","timestamp":1722817424798,"payload":"something went wrong","tid":3472525}
+///    ```
+/// 
+///  - If the logger has a name:
+/// 
+///    ```json
+///    {"level":"Info","timestamp":1722817541459,"payload":"hello, world!","logger":"app-component","tid":3478045}
+///    {"level":"Error","timestamp":1722817541459,"payload":"something went wrong","logger":"app-component","tid":3478045}
+///    ```
+/// 
+///  - If crate feature `source-location` is enabled:
+/// 
+///    ```json
+///    {"level":"Info","timestamp":1722817572709,"payload":"hello, world!","tid":3479856,"source":{"module_path":"my_app::say_hi","file":"src/say_hi.rs","line":4,"column":5}}
+///    {"level":"Error","timestamp":1722817572709,"payload":"something went wrong","tid":3479856,"source":{"module_path":"my_app::say_hi","file":"src/say_hi.rs","line":5,"column":5}}
+///    ```
+/// 
+/// [`Level::as_str`]: crate::Level::as_str
+/// [`SourceLocation`]: crate::SourceLocation
+#[derive(Clone)]
+pub struct JsonFormatter(PhantomData<()>);
+
+impl JsonFormatter {
+    /// Constructs a `JsonFormatter`.
+    #[must_use]
+    pub fn new() -> JsonFormatter {
+        JsonFormatter(PhantomData)
+    }
+
+    fn format_impl(
+        &self,
+        record: &Record,
+        dest: &mut StringBuf,
+    ) -> Result<FmtExtraInfo, JsonFormatterError> {
+        cfg_if! {
+            if #[cfg(not(feature = "flexible-string"))] {
+                dest.reserve(crate::string_buf::RESERVE_SIZE);
+            }
+        }
+
+        let json_record: JsonRecord = record.into();
+
+        // TODO: https://github.com/serde-rs/json/issues/863
+        //
+        // The performance can be significantly optimized here if the issue can be
+        // solved.
+        dest.write_str(&serde_json::to_string(&json_record)?)?;
```

Comment:
Leaving this comment as a note. 

```
test bench_1_full_formatter       ... bench:          90.39 ns/iter (+/- 9.58)
test bench_1_json_formatter       ... bench:         121.27 ns/iter (+/- 9.10)
```

I benchmarked `JsonFormatter` and I was expecting it to be faster than that (close to `FullFormatter`). After a quick check, the `String` returned by `serde::to_string` will do unnecessary allocation and should be optimized out.

Unfortunately, due to the `serde_json` API limitation (serde-rs/json#863), this is currently not possible (at least not without using unsafe).